packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol

```diff
@@ -426,70 +433,66 @@ contract PerpetualLiquidatable is PerpetualPositionManager {
             // Pay LIQUIDATOR: collateral + dispute bond + returned final fee
-            withdrawalAmount = collateral.add(disputeBondAmount).add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(disputeBondAmount).add(finalFee);
+
             // If the state is pre-dispute but time has passed liveness then there was no dispute. We represent this
             // state as a dispute failed and the liquidator can withdraw.
-        } else if (liquidation.state == Status.PreDispute && msg.sender == liquidation.liquidator) {
+        } else if (liquidation.state == Status.PreDispute) {
             // Pay LIQUIDATOR: collateral + returned final fee
-            withdrawalAmount = collateral.add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(finalFee);
         }
-        require(withdrawalAmount.isGreaterThan(0), "Invalid withdrawal amount");
 
-        // Decrease the total collateral held in liquidatable by the amount withdrawn.
-        amountWithdrawn = _removeCollateral(rawLiquidationCollateral, withdrawalAmount);
+        require(
+            (rewards.payToLiquidator.add(rewards.payToSponsor).add(rewards.payToDisputer)).isGreaterThan(0),
+            "Nothing to withdraw"
+        );
 
-        emit LiquidationWithdrawn(msg.sender, amountWithdrawn.rawValue, liquidation.state, settlementPrice.rawValue);
+        // Decrease the total collateral held in liquidatable by the amount to pay each party. The actual amounts
+        // withdrawn might differ if _removeCollateral causes precision loss.
+        if (rewards.payToLiquidator.isGreaterThan(0)) {
+            rewards.paidToLiquidator = _removeCollateral(rawLiquidationCollateral, rewards.payToLiquidator);
+            collateralCurrency.safeTransfer(liquidation.liquidator, rewards.paidToLiquidator.rawValue);
+        }
+        if (rewards.payToSponsor.isGreaterThan(0)) {
+            rewards.paidToSponsor = _removeCollateral(rawLiquidationCollateral, rewards.payToSponsor);
+            collateralCurrency.safeTransfer(liquidation.sponsor, rewards.paidToSponsor.rawValue);
+        }
+        if (rewards.payToDisputer.isGreaterThan(0)) {
+            rewards.paidToDisputer = _removeCollateral(rawLiquidationCollateral, rewards.payToDisputer);
+            collateralCurrency.safeTransfer(liquidation.disputer, rewards.paidToDisputer.rawValue);
+        }
 
-        // Transfer amount withdrawn from this contract to the caller.
-        collateralCurrency.safeTransfer(msg.sender, amountWithdrawn.rawValue);
+        emit LiquidationWithdrawn(
+            msg.sender,
+            rewards.paidToLiquidator.rawValue,
+            rewards.paidToDisputer.rawValue,
+            rewards.paidToSponsor.rawValue,
+            liquidation.state,
+            settlementPrice.rawValue
+        );
 
-        return amountWithdrawn;
```

Comment:
do we want to return anything? it could be useful to return 3 uints for the amount sent to the liquidator, disputer and sponsor. WDYT?

Reply:
However, this will increase bytecode a bit and I'm not sure how useful it is. Just thinking from a composability perspective: if you were connecting the EMP to another contract that preforms liquidations you'd want to be able to see these numbers returned from the function call.

Reply:
> do we want to return anything? it could be useful to return 3 uints for the amount sent to the liquidator, disputer and sponsor. WDYT?

It could yeah. Currently the bots use this return value to check if they are getting > 0 rewards so maybe we return the 3 numbers. It doesn't use much bytecode, ~200 bytes

Reply:
Addressed in this [commit](https://github.com/UMAprotocol/protocol/pull/2152/commits/ba30b32ddad538de6e07dadfff09e7ace7e8989f)